contracts/SimpleGAMultiSig.aes

```diff
@@ -22,36 +22,46 @@
 @compiler >= 6
 
 include "Option.aes"
+include "Set.aes"
 include "String.aes"
 
-// This contract provides multi signature based on generalized accounts
-// A transaction can be proposed by one of the signers which are registered during contract deployment
-// The transaction must be confirmed by state.confirmations_required of the registered signers, while the proposing user automatically confirms
-// If the amount of confirmations is reached, authenticate can be called by any user and the original tx is processed
-// Only one proposed transaction can be multisigned at a time
+// This contract provides multi signature based on generalized accounts.
+// A transaction can be proposed by one of the signers which are registered during contract deployment.
+// The transaction must be confirmed by a certain amount of signers. This is also defined during contract deployment.
+// Proposing a transaction automatically counts as confirmation.
+// When proposing a tx, the signer can define a TTL until the proposed tx needs to be confirmed and authorized.
+// If the amount of confirmations is reached, the special function "authorize" can be called by anybody.
+// Only one transaction can be handled at the same time.
+// Every signer can revoke the proposed transaction at any time (no matter if it's expired or not).
 contract SimpleGAMultiSig =
 
   datatype event
     = TxProposed(hash, address, int)
     | TxConfirmed(hash, address)
     | TxConsensusReached(hash)
+    | TxRevoked(hash, address)
     | TxAuthorized(hash)
-    | TxRevoked(address)
 
-  record transaction = 
+  record transaction =
     { ga_tx_hash : hash,
       expires_at : int,
       proposed_by : address,
-      confirmed_by : list(address) }
+      confirmed_by : Set.set(address) }
 
-  record state = 
-  	{ signers : list(address),
-  	  confirmations_required : int,    
+  record consensus_info =
+    { ga_tx_hash : option(hash),
+      confirmations_required : int,
+      confirmed_by : list(address),
+      can_authorize : bool }
+
+  record state =
+  	{ signers : Set.set(address),
+  	  confirmations_required : int,
       current_tx : option(transaction),
       nonce : int }
 
   stateful entrypoint init(p_confirmations_required : int, p_cosigners : list(address)) =
-    { signers = Call.caller :: p_cosigners,
+    { signers = Set.insert(Call.caller, Set.from_list(p_cosigners)),
```

Comment:
set is good to use, but `from_list` should be avoided

Reply:
what's the problem here? is this relevant for gas consumption? I definitely prefer a convenient way to provide the list of co-signers. passing this in the params as ´map` seems to be inconvenient to me

Reply:
from_list iterates, to it consumes more gas per entry, passing map is the way to go

Reply:
Converting from list to Set means a bit more gas used (for normal usage with a handful of co-signers it won't be a lot though). The alternative is to build a set (i.e. mapping addresses to `unit`) and that is an implementation detail that perhaps is best hidden from the user. 🤔 

Fun fact: The map will be marginally larger (and we pay for the byte size of the calldata) - but converting from a list is still more expensive.

Reply:
I am still wrapping my head around this. as we use a `Set.set` for internal logic after initialization I don't see that much of a problem here. I know it's just convenience for the param in the `init` entrypoint. but I still tend to favor using `list` here given the fact that we will probably have only a handful cosigners included in it.

Reply:
I am postponing this now and opened an issue, see https://github.com/aeternity/ga-multisig-contract/issues/3